writer.go

```diff
@@ -72,7 +72,11 @@ func (w *Writer) Flush() error {
 		}
 		return err
 	}
-	w.buf = w.buf[n:]
+	// io.Writer.Write contract says n == len(w.buf) if err == nil, so copy below *should* be a no-op.
+
+	copy(w.buf, w.buf[n:]) // compact buffer
+	w.buf = w.buf[:0]      // recycle buffer memory
```

Comment:
Having both of these is redundant. The first moves non-written data to the front, the other clears the buffer. If you wish to clear the buffer like this, you would need to check for a short write:
```
n, err := w.out.Write(w.buf)
if n < len(w.bif) {
  w.Error = io.ErrShortWrite
  return w.Error
}
// other error handling...
w.buf = w.buf[ : 0]
```

Reply:
Oh, you're absolutely right 🤦 . I think it's best to replace `w.buf[:0]` right `w.buf[:n]` to preserve the copied data, rather than error out here. It is a violation of `io.Writer` contract to short-write but I imagine there are many real writers violate this...

Reply:
Alternatively I could add a loop to keep writing until it's drained. Do you have a preference?

Reply:
Ah, cool :+1: